OsmAnd-java/src/main/java/net/osmand/router/RouteColorize.java

```diff
@@ -0,0 +1,416 @@
+package net.osmand.router;
+
+import net.osmand.GPXUtilities;
+import net.osmand.osm.edit.Node;
+import net.osmand.osm.edit.OsmMapUtils;
+import net.osmand.util.MapUtils;
+
+import java.awt.Color;
+import java.util.ArrayList;
+import java.util.List;
+
+public class RouteColorize {
+
+    public int zoom;
+    public double[] latitudes;
+    public double[] longitudes;
+    public double[] values;
+    public double minValue;
+    public double maxValue;
+    public double[][] palette;
+
+    private List<Data> dataList;
+
+    public static final int DARK_GREY = -10724260;
+    public static final int LIGHT_GREY = -3618616;
+    public static final int RED = -65279;
+    public static final int GREEN = -1087457024;
+    public static final int YELLOW = -469770750;
+
+    public enum ValueType {
+        ELEVATION,
+        SPEED,
+        SLOPE,
+        NONE
+    }
+
+    private final int VALUE_INDEX = 0;
+    private final int DECIMAL_COLOR_INDEX = 1;//sRGB decimal format
+    private final int RED_COLOR_INDEX = 1;//RGB
+    private final int GREEN_COLOR_INDEX = 2;//RGB
+    private final int BLUE_COLOR_INDEX = 3;//RGB
+    private final int ALPHA_COLOR_INDEX = 4;//RGBA
+
+    private ValueType valueType;
+
+    public static int SLOPE_RANGE = 150;
+
+
+    /**
+     * @param minValue can be NaN
+     * @param maxValue can be NaN
+     * @param palette  array {{value,color},...} - color in sRGB (decimal) format OR {{value,RED,GREEN,BLUE,ALPHA},...} - color in RGBA format
+     */
+    public RouteColorize(int zoom, double[] latitudes, double[] longitudes, double[] values, double minValue, double maxValue, double[][] palette) {
+        this.zoom = zoom;
+        this.latitudes = latitudes;
+        this.longitudes = longitudes;
+        this.values = values;
+        this.minValue = minValue;
+        this.maxValue = maxValue;
+        this.palette = palette;
+
+        if (Double.isNaN(minValue) || Double.isNaN(maxValue)) {
+            calculateMinMaxValue();
+        }
+        checkPalette();
+        sortPalette();
+    }
+
+    /**
+     * @param type ELEVATION, SPEED, SLOPE
+     */
+    public RouteColorize(int zoom, GPXUtilities.GPXFile gpxFile, ValueType type) {
+
+        if (!gpxFile.hasTrkPt())
+            return;
+
+        List<Double> latList = new ArrayList<>();
```

Comment:
we use TDoubleArrayList, so conversion is better between Double and double

Reply:
For use TDoubleArrayList need add library in Gradle